fs-index/src/watch.rs

```diff
@@ -0,0 +1,69 @@
+use crate::{ResourceIndex, ARK_FOLDER};
+use anyhow::Result;
+use futures::{
+    channel::mpsc::{channel, Receiver},
+    SinkExt, StreamExt,
+};
+use log::info;
+use notify::{Config, Event, RecommendedWatcher, RecursiveMode, Watcher};
+use std::{fs, path::Path};
+
+/// Watch the root path for changes and update the index
+pub async fn watch_index<P: AsRef<Path>>(root_path: P) -> Result<()> {
+    log::debug!(
+        "Attempting to watch index at root path: {:?}",
+        root_path.as_ref()
+    );
+
+    let root_path = fs::canonicalize(root_path.as_ref())?;
+    let mut index = ResourceIndex::provide(&root_path)?;
+
+    let (mut watcher, mut rx) = async_watcher()?;
+    info!("Watching directory: {:?}", root_path);
+    let config = Config::default();
+    watcher.configure(config)?;
+    watcher.watch(root_path.as_ref(), RecursiveMode::Recursive)?;
+    info!("Started watcher with config: \n\t{:?}", config);
+
+    while let Some(res) = rx.next().await {
+        match res {
+            Ok(event) => {
+                // If the event is a change in .ark folder, ignore it
+                let ark_folder = root_path.join(ARK_FOLDER);
+                if event
+                    .paths
+                    .iter()
+                    .any(|p| p.starts_with(&ark_folder))
+                {
+                    continue;
+                }
+
+                info!("Detected event: {:?}", event);
+                index.update_all()?;
```

Comment:
This looks too inefficient. If we call `update_all` for each update, why not just call it once in a while? I think we should take changed paths, filter out those starting with `.ark`, then update index only using the modified paths.

Reply:
You're right. 
I went through the docs for the event types
```rust
pub enum EventKind {
    Any,
    Access(AccessKind),
    Create(CreateKind),
    Modify(ModifyKind),
    Remove(RemoveKind),
    Other,
}
```

I suggest pausing the changes in this PR until we update the ResourceIndex API as discussed in [https://github.com/ARK-Builders/arklib/issues/91#issuecomment-1998887042](https://github.com/ARK-Builders/arklib/issues/91#issuecomment-1998887042)

Then, we can follow this workflow:
- Check for `need_rescan()`. If true, call `update_all()`.
- Filter out events in `.ark/`.
- Use a match statement for the type of event:
    - `Create` -> call `track_addition` with the relative path.
    - `Modify` -> call `track_update`.
    - `Remove` -> call `track_deletion`.

These actions seem to align perfectly with the methods we agreed on for the API.

Reply:
Cool that the "Track API" appeared to be useful already.

By the way, we might utilize `Access(AccessKind)`, too. Like providing "recently viewed" feature, which would take into account all users devices. We should design this feature later.